src/hotspot/share/runtime/objectMonitor.cpp

```diff
@@ -312,7 +312,60 @@ void ObjectMonitor::ClearSuccOnSuspend::operator()(JavaThread* current) {
 // -----------------------------------------------------------------------------
 // Enter support
 
+bool ObjectMonitor::enter_for(JavaThread* locking_thread) {
+  // Used by ObjectSynchronizer::enter_for to enter for another thread
+  // this code may only contend with deflation.
+  assert(locking_thread == Thread::current() || locking_thread->is_obj_deopt_suspend(), "must be");
+
+  // Block out deflation as soon as possible.
+  add_to_contentions(1);
+
+  bool success = false;
+  if (!is_being_async_deflated()) {
+    void* prev_owner = try_set_owner_from(nullptr, locking_thread);
+
+    if (prev_owner == nullptr) {
+      assert(_recursions == 0, "invariant");
+      success = true;
+    } else if (prev_owner == locking_thread) {
+      _recursions++;
+      success = true;
+    } else if (prev_owner == DEFLATER_MARKER) {
+      // Racing with deflation.
+      prev_owner = try_set_owner_from(DEFLATER_MARKER, locking_thread);
+      if (prev_owner == DEFLATER_MARKER) {
+        // Cancelled deflation. Increment contentions as part of the deflation protocol.
+        add_to_contentions(1);
+        success = true;
+      } else if (prev_owner == nullptr) {
+        // At this point we cannot race with deflation as we have both incremented
+        // contentions, seen contention > 0 and seen a DEFLATER_MARKER.
+        // success will only be false if this races with something other than
+        // deflation.
+        success = try_set_owner_from(nullptr, locking_thread) == nullptr;
```

Comment:
No other race than the race with deflation should be allowed because the current thread would be busy waiting. Better assert success?

Reply:
I was unsure if the `ObjectMonitor::enter_for` should assert this or put it on the caller. 
`ObjectSynchronizer::enter_for` currently asserts `assert(monitor->is_being_async_deflated(), "must be");` when  `ObjectMonitor::enter_for` return false. (An equivalent property)

But I can move it into `ObjectMonitor::enter_for` and attach some more information to the assert such as the prev_owner. Sounds better to me.

Reply:
Sorry, I overlooked the other assertion.